### ISSUE TITLE
Improved boxing based on usage

### ISSUE THREAD

**Issue by tombentley** [ceylon/ceylon-compiler#1464]
Consider:

``` ceylon
class BugXXXX(Iterable<Integer> ints) {
    for (Integer int in ints) {
        print(int);
        print(int);
    }
}
```

Ignoring the optimizations done in #882 for a moment, the loop gets transformed to something like:

``` java
        .java.lang.Object elem$0;
        .ceylon.language.Iterator<? extends .ceylon.language.Integer> int$iterator$$1 = ints.iterator();
        while (!((elem$0 = int$iterator$$1.next()) instanceof .ceylon.language.Finished)) {
            final long int = ((.ceylon.language.Integer)elem$0).longValue();
            .ceylon.language.print_.print(.ceylon.language.Integer.instance(int));
            .ceylon.language.print_.print(.ceylon.language.Integer.instance(int));
        }
```

So although `int` is used _exclusively_ with boxed APIs within the body of the loop we none the less unbox it, and then have to box it again for _each_ call. Obviously this generates a lot of garbage.

It would be nice if the `BoxingVisitor` could look at how the loop variable is used within the body of the loop and convey this information to the loop transformations so that:
1. We only unbox the loop variable when we have to (i.e. when its used with APIs that take unboxed arguments)
2. We use the boxed instance obtained from the iterator and don't go boxing the unboxed variable for every invocation.

Obviously in the case of the loop optimizations done in #882, the transformed loop may naturally yield unboxed variables, in which case we want to box the variable once.

**Comment by quintesse** [comment]
Ok, I just pushed the final fixes to the current boxing code as a first step to handle this issue.
What I added for now is the system to change the boxing of declarations, mainly meant for locals but not really limited to those. What you can do is:

 - Add `-Dceylon.compile.forceBoxedLocals=true` somewhere on the command line or in your IDE settings. This will cause **all** locals to be generated as boxed instead of the default unboxed state.
 - Add `@boxed` or `@unboxed` compiler annotations to force specific declarations to be either boxed or unboxed.

In the latter case the `@unboxed` doesn't make sense right now because thats the default. But when this issue gets closed we will have some heuristics that might make certain locals boxed, so with `@unboxed` you could override that again.

Also, the `-Dceylon.compile.forceBoxedLocals` only changes locals, but the `@boxed` / `@unboxed` annotations can be used for all kinds of declarations like attributes and methods, either local or member (didn't test toplevel).

And finally, these are for *testing purposes*! It might be nice to play around with them but code generation might fail. Or you could get nonsensical situations. I only validated this for locals.

**Comment by FroMage** [comment]
Is this going to make it in 1.2?

**Comment by quintesse** [comment]
Yes, we can at least put in an initial implementation with simple heuristics and expand on them when we encounter cases that we want improved.
I currently have an implementation that improves the situation in some cases and worsens in others. If I can at least prevent those latter cases I can merge what I have.

**Comment by tombentley** [comment]
@quintesse any progress on this?

**Comment by quintesse** [comment]
It's very close, but I'm not making too much progress on the last issue I'm having.
The biggest part of the work has been done at least: we can now override at will if a declaration should be treated as boxed or unboxed. (That's all pushed)
And I have a visitor that looks at how locals are used and can decide to make them boxed instead of unboxed (which is what we do right now in all cases). Currently it's pretty simple , it just says that if there's *any* usage where a boxed value is needed it just makes the local boxed (the idea being that unboxing is always cheaper than boxing).
Problem is that in certain cases (having to do with operator methods like `plus()`) it makes the wrong decision and makes things worse. I'm trying to figure out how to fix that.

**Comment by quintesse** [comment]
I stepped away from it for a while. I'll see if I can close it today or move it to 1.3 otherwise.

**Comment by FroMage** [comment]
Let's move it to 1.3
